### PR TITLE
Roller Chairs

### LINKED ISSUE #2425: Make Roller Chairs draggable B: 50
# Criteria:

- Make roller chairs draggable and also play the chair rolling sfx periodically
- Change the direction of the chair so it faces the direction of dragging
- Allow players buckled to the chair to be dragged along and also face the same direction as the chair being dragged
- Ensure all of it works over multiplayer with 2 clients

Contributor will receive a bounty of 50usd on a successful merge of a solution that resolves the task request above.

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Fixes #2425 
You can even do newtonian movement while seated in space by picking up a gun or something and shooting. Doesn't work while there's gravity though.

This was half and half between me and @JesterX666 . You can give them half and put my half back into the pool so we can put it towards other bounties.

### Notes:
There might be some weird edge cases, particularly when in space, but situations I thought to test seem to be working fine.

### Please make sure you have followed the self checks below before submitting a PR:

- Code is sufficiently commented
- Code is indented with tabs and not spaces
- The PR does not include any unnecessary .meta, .prefab or <b>.unity (scene) changes</b>
- The PR does not bring up any new compile errors
- The PR has been tested in editor
- Any new files are named using PascalCase (to avoid issues on case sensitive file systems)
- Any new / changed components follow the [Component Development Checklist](https://github.com/unitystation/unitystation/wiki/Component-Development-Checklist)
- Any new objects / items follow the [Creating Items and Objects Guide](https://github.com/unitystation/unitystation/wiki/Creating-Items-and-Objects%3A-Now-With-Prefab-Variants) (especially concerning the use of prefab variants)
- The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- The PR has been tested with round restarts.
- The PR has been tested on moving / rotating / rotated-before-joining matrices (if applicable)

### CODE REVIEW

File: UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs [hunk 1]
```diff
@@ -75,7 +84,7 @@ public void Awake()
 	private void OnWillDestroyServer(DestructionInfo info)
 	{
 		//release the player
-		if (occupant > 0)
+		if (occupantNetId > 0)
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs]: Will be preferable to use the NetId.Empty constant instead of 0 (makes it more pretty)

File: UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs [hunk 2]
```diff
@@ -119,7 +128,7 @@ private void UpdateFrontSprite()
 	{
 		if (spriteRendererFront)
 		{
-			if (occupant > 0)
+			if (occupantNetId > 0)
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs]: Will be preferable to use the NetId.Empty constant instead of 0 (makes it more pretty)

File: UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs [hunk 3]
```diff
@@ -170,7 +187,7 @@ private void SyncOccupant(uint occupantNewValue)
 	//ensures we are rendering in the correct sprite layer
 	private void EnsureSpriteLayer()
 	{
-		if (directional.CurrentDirection == Orientation.Up && occupant > 0)
+		if (directional.CurrentDirection == Orientation.Up && occupantNetId > 0)
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs]: Will be preferable to use the NetId.Empty constant instead of 0 (makes it more pretty)

File: UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs (hunk 1 above)
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Furniture/OccupiableDirectionalSprite.cs]: Will be preferable to use the NetId.Empty constant instead of 0 (makes it more pretty)

File: UnityProject/Assets/Scripts/Networking/NetworkUtils.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/// </summary>
+public static class NetworkUtils
+{
+	/// <summary>
+	/// Tries to find the object with the given net ID, returns null if unsuccessful. Logs
+	/// a warning only if the netID is not empty and not invalid but not found.
+	/// </summary>
+	/// <param name="netId"></param>
+	/// <returns></returns>
+	public static GameObject FindObjectOrNull(uint netId)
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Networking/NetworkUtils.cs:16]: Nice and useful method that centralize a very frequent call including a log is a good idea.

File: UnityProject/Assets/Scripts/Player/PlayerMove.cs [hunk 5]
```diff
@@ -21,16 +21,23 @@ public class PlayerMove : NetworkBehaviour, IRightClickable, IServerSpawn
 	[SyncVar] public bool allowInput = true;
 
 	//netid of the game object we are buckled to, NetId.Invalid if not buckled
-	[SyncVar(hook = nameof(OnBuckledChangedHook))]
-	public uint buckledObject = NetId.Invalid;
+	[SyncVar(hook = nameof(SyncBuckledObjectNetId))]
+	private uint buckledObjectNetId = NetId.Invalid;
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Player/PlayerMove.cs]: I don't know if NetId.Empty would make more sense than NetId.Invalid to represent a "not buckled" state?

File: UnityProject/Assets/Scripts/Player/PlayerMove.cs [hunk 6]
```diff
@@ -373,12 +380,28 @@ public void CmdUnbuckle()
...
 		//decide if we should fall back down when unbuckled
 		registerPlayer.ServerSetIsStanding(PlayerScript.playerHealth.ConsciousState == ConsciousState.CONSCIOUS);
 		onUnbuckled?.Invoke();
+		if (previouslyBuckledTo)
+		{
+			//we are unbuckled but still will drift with the object.
+			var buckledCNT = previouslyBuckledTo.GetComponent<CustomNetTransform>();
+			if (buckledCNT.IsFloatingServer)
+			{
+				playerScript.PlayerSync.NewtonianMove(buckledCNT.ServerImpulse.NormalizeToInt(), buckledCNT.SpeedServer);
```
Comment (JesterX666) [inline UnityProject/Assets/Scripts/Player/PlayerMove.cs:396]: Cool  :)

**Review by JesterX666 — COMMENTED** [review]
Can't wait to try it :)

**Comment by chairbender** [comment]
Applied the suggestions.
